File: Codes in Python3/Binary Tree/Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        curr=root
        res=[]
        while curr:
            if curr.left is None:
                res.append(curr.val)
                curr=curr.right
            else:
                prev=curr.left
                while prev.right and prev.right!=curr:
                    prev=prev.right
                if prev.right is None:
                    prev.right=curr
                    curr=curr.left
                else:
                    prev.right=None
                    res.append(curr.val)
                    curr=curr.right
        return res
        # stack=[]
        # res=[]
        # while root or stack:
        #     while root:
        #         stack.append(root)
        #         root=root.left
        #     root=stack.pop()
        #     res.append(root.val)
        #     root=root.right
        # return res
        # return self.inorderTraversal(root.left) + [root.val] + self.inorderTraversal(root.right) if root else []

File: Codes in C++/Binary Tree/Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<int> inorderTraversal(TreeNode *root)
    {
        vector<int> inorder;
        TreeNode *curr = root;
        while (curr != NULL)
        {
            if (curr->left == NULL)
            {
                inorder.push_back(curr->val);
                curr = curr->right;
            }
            else
            {
                TreeNode *prev = curr->left;
                while (prev->right && prev->right != curr)
                    prev = prev->right;
                if (prev->right == NULL)
                {
                    prev->right = curr;
                    curr = curr->left;
                }
                else
                {
                    prev->right = NULL;
                    inorder.push_back(curr->val);
                    curr = curr->right;
                }
            }
        }
        return inorder;
    }
    /* vector<int> inorderTraversal(TreeNode* root) {
     vector<int> res;
     stack<TreeNode*> s;
     while(root!=NULL||!s.empty())
     {
         while(root!=NULL)
         {
             s.push(root);
             root=root->left;
         }
         root=s.top();
         s.pop();
         res.push_back(root->val);
         root=root->right;
     }
     return res;
     }*/
    /*public:
        vector<int> inorderTraversal(TreeNode* root) {
            vector<int> nodes;
            inorder(root, nodes);
            return nodes;
        }
    private:
        void inorder(TreeNode* root, vector<int>& nodes) {
            if (!root) {
                return;
            }
            inorder(root -> left, nodes);
            nodes.push_back(root -> val);
            inorder(root -> right, nodes);
        }*/
};